src/chain/store/index.rs

```diff
@@ -151,139 +62,186 @@ impl<DB: Blockstore> ChainIndex<DB> {
         Ok(ts)
     }
 
-    /// Loads tipset at `to` [`ChainEpoch`], loading from sparse cache and/or
-    /// loading parents from the `blockstore`.
-    pub(in crate::chain) fn get_tipset_by_height(
+    /// Find tipset at epoch `to` in the chain indicated by `from`. The `from`
+    /// tipset's epoch must not be smaller than `to`.
+    ///
+    /// # Why pass in the `from` argument?
+    ///
+    /// Imagine the database contains five tipsets and a genesis block in this
+    /// configuration:
+    ///
+    /// ```text
+    ///           ┌───────┐  ┌────────┐  ┌────────┐
+    /// Genesis◄──┤Epoch 1◄──┤Epoch 2A◄──┤Epoch 3A│
+    ///           └───▲───┘  └────────┘  └────────┘
+    ///               │      ┌────────┐  ┌────────┐
+    ///               └──────┤Epoch 2B◄──┤Epoch 3B│
+    ///                      └────────┘  └────────┘
+    /// ```
+    ///
+    /// Here we have a fork in the chain and it is ambiguous which tipset to
+    /// load when epoch 2 is requested. The ambiguity is solved by passing in a
+    /// younger tipset (higher epoch) from which has the desired tipset as an
+    /// ancestor.
+    /// Calling `get_tipset_by_height(2, epoch_3a)` will return `Epoch 2A`.
+    /// Calling `get_tipset_by_height(2, epoch_3b)` will return `Epoch 2B`.
+    ///
+    /// # What happens when a null tipset is requested?
+    ///
+    /// ```text
+    ///           ┌───────┐          ┌───────┐  ┌───────┐
+    /// Genesis◄──┤Epoch 1│   Null   │Epoch 3◄──┤Epoch 4│
+    ///           └───▲───┘          └───┬───┘  └───────┘
+    ///               │                  │
+    ///               └──────────────────┘
+    /// ```
+    /// If the requested epoch points to a null tipset, there are two options:
+    /// Pick the nearest older tipset or pick the nearest younger tipset.
+    /// Requesting epoch 2 with `ResolveNullTipset::TakeYounger` will return
+    /// epoch 3. Requesting with `ResolveNullTipset::TakeOlder` will return
+    /// epoch 1.
+    pub fn tipset_by_height(
         &self,
-        from: Arc<Tipset>,
         to: ChainEpoch,
+        from: Arc<Tipset>,
+        resolve: ResolveNullTipset,
     ) -> Result<Arc<Tipset>, Error> {
-        if from.epoch() - to <= SKIP_LENGTH {
-            return self.walk_back(from, to);
+        if to == 0 {
+            return Ok(Arc::new(Tipset::from(from.genesis(&self.db)?)));
+        }
+        if to > from.epoch() {
+            return Err(Error::Other(
+                "Looking for tipset with height greater than start point".to_string(),
+            ));
```

Comment:
> The `from` tipset's epoch must not be smaller than `to`.

Function contract is unclear about the consequences. Should we panic in this case? I think so. Otherwise document this as a trivial failure